File: message_input/src/main/java/com/emamagic/message_input/MessageEditText.java
package com.emamagic.message_input;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;

import com.emamagic.emoji.view.EmojiMultiAutoCompleteTextView;


public class MessageEditText extends EmojiMultiAutoCompleteTextView {

    private KeyPreImeListener keyPreImeListener;

    public MessageEditText(Context context) {
        super(context);
        init();
    }

    public MessageEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public MessageEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    @Override
    public boolean onKeyPreIme(int keyCode, KeyEvent event) {
        if (keyPreImeListener != null
                && keyCode == KeyEvent.KEYCODE_BACK) {
            return keyPreImeListener.onHideKeyboard();
        }
        return super.onKeyPreIme(keyCode, event);
    }

    public void setKeyPreImeListener(KeyPreImeListener keyPreImeListener) {
        this.keyPreImeListener = keyPreImeListener;
    }

    private void init() {
        if (isInEditMode()) {
            return;
        }
        setInputType(getInputType() | EditorInfo.IME_FLAG_NO_EXTRACT_UI | EditorInfo.IME_FLAG_NO_FULLSCREEN);
        setImeOptions(getImeOptions() | EditorInfo.IME_FLAG_NO_FULLSCREEN);
    }

    @SuppressLint("SetTextI18n")
    public void setText(CharSequence text, TextView.BufferType type) {
        try {
            super.setText(text, type);
        } catch (Exception e) {
            setText("I tried, but your OEM just sucks because they modify the framework components and therefore causing the app to crash!" + ".\nFastHub");
        }
    }

    public interface KeyPreImeListener {

        /**
         * Fires when KEYCODE_BACK key pressed
         */
        boolean onHideKeyboard();
    }
}

File: chat_box/src/main/java/com/emamagic/chat_box/MessageEditText.kt
package com.emamagic.chat_box

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.inputmethod.EditorInfo
import com.emamagic.emoji.view.EmojiMultiAutoCompleteTextView
import java.lang.Exception

class MessageEditText @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null
) : EmojiMultiAutoCompleteTextView(context, attrs) {


    init { init() }

    private var keyPreImeListener: KeyPreImeListener? = null

    override fun onKeyPreIme(keyCode: Int, event: KeyEvent?): Boolean {
        return if (keyPreImeListener != null
            && keyCode == KeyEvent.KEYCODE_BACK
        ) {
            keyPreImeListener!!.onHideKeyboard()
        } else super.onKeyPreIme(keyCode, event)
    }

    fun setKeyPreImeListener(keyPreImeListener: KeyPreImeListener?) {
        this.keyPreImeListener = keyPreImeListener
    }

    private fun init() {
        if (isInEditMode) {
            return
        }
        inputType =
            inputType or EditorInfo.IME_FLAG_NO_EXTRACT_UI or EditorInfo.IME_FLAG_NO_FULLSCREEN
        imeOptions = imeOptions or EditorInfo.IME_FLAG_NO_FULLSCREEN
    }

    @SuppressLint("SetTextI18n")
    override fun setText(text: CharSequence?, type: BufferType?) {
        try {
            super.setText(text, type)
        } catch (e: Exception) {
            setText(
                """
                I tried, but your OEM just sucks because they modify the framework components and therefore causing the app to crash!.
                FastHub
                """.trimIndent()
            )
        }
    }

    interface KeyPreImeListener {
        fun onHideKeyboard(): Boolean
    }
}